crates/sui-framework/sources/test_scenario_wip.move

```diff
@@ -0,0 +1,305 @@
+
+#[test_only]
+module sui::test_scenario_wip {
+    use sui::object::{Self, ID, UID};
+    use sui::tx_context::{Self, TxContext};
+    use std::option::{Self, Option};
+
+    /// Asserted transaction result was a success, but it was not
+    const ETransactionNotSuccessful: u64 = 0;
+
+    /// Asserted transaction result was not successful, but it was
+    const ETransactionSuccessful: u64 = 1;
+
+    /// Transaction ended without all shared and immutable objects being returned
+    const EObjectsNotReturned: u64 = 2;
+
+    /// Attempted to return an object to the inventory that was not previously removed from the
+    /// inventory during the current transaction. Can happen if the user attempts to call
+    /// `return_to_address` on a locally constructed object rather than one returned from a
+    /// `test_scenario` function such as `take_from_address`.
+    const ECantReturnObject: u64 = 3;
+
+    /// Attempted to retrieve an object of a particular type from the inventory, but it is empty.
+    /// Can happen if the user already transferred the object or a previous transaction failed to
+    /// transfer the object to the user.
+    const EEmptyInventory: u64 = 4;
+
+    /// Object of that ID was not found in that inventory. It was possibly already taken
+    const EObjectNotFound: u64 = 5;
+
+    /// Object was found, but the wrong type was specified.
+    const EWrongObjectType: u64 = 6;
+
+    /// Utility for mocking a multi-transaction Sui execution in a single Move procedure.
+    /// A `Scenario` maintains a view of the global object pool built up by the execution.
+    /// These objects can be accessed via functions like `take_from_sender`, which gives the
+    /// transaction sender access to objects in (only) their inventory.
+    /// Example usage:
+    /// ```
+    /// let addr1: address = 0;
+    /// let addr2: address = 1;
+    /// // begin a test scenario in a context where addr1 is the sender
+    /// let scenario = &mut test_scenario::begin(&addr1);
+    /// // addr1 sends an object to addr2
+    /// {
+    ///     let some_object: SomeObject = ... // construct an object
+    ///     transfer::transfer(some_object, copy addr2)
+    /// };
+    /// // end the first transaction and begin a new one where addr2 is the sender
+    /// test_scenario::assert_success(test_scenario::next_tx(scenario, &addr2));
+    /// {
+    ///     // remove the SomeObject value from addr2's inventory
+    ///     let obj = test_scenario::take_from_sender<SomeObject>(scenario);
+    ///     // use it to test some function that needs this value
+    ///     SomeObject::some_function(obj)
+    /// }
+    /// ... // more txes
+    /// ```
+    struct Scenario {
+        txn_number: u64,
+        ctx: TxContext,
+    }
+
+    struct TransactionResult {
+        /// Some if the "transaction" was successful
+        /// None otherwise
+        effects: Option<TransactionEffects>,
+    }
+
+    struct TransactionEffects has drop {
```

Comment:
It might be useful to have a separate `transferred` field here (even though transfers are technically a subset of writes) + include something like an `option<address>` to say where it was transferred to.